odoo/models.py

```diff
@@ -4240,6 +4244,17 @@ def _load_records(self, data_list, update=False):
                 if data.get('xml_id') and not data['xml_id'].startswith(prefix):
                     _logger.warning("Creating record %s in module %s.", data['xml_id'], module)
 
+        if self.env.context.get('import_file'):
+            existing_modules = self.env['ir.module.module'].search([]).mapped('name')
+            for data in to_create:
+                xml_id = data.get('xml_id')
+                if xml_id:
+                    module_name = xml_id.split('.')[0]
+                    if module_name in existing_modules:
+                        raise UserError(
+                            _("The record %(xml_id)s with module prefix %(module_name)s will be deleted when the module is upgraded. Use either no prefix or a prefix that isn't an existing module.",
```

Comment:
I am not sure if `no prefix` for xmlid is legal. Because we have many codes like `module = xmlid.split('.', 1)[0]`, no prefix xmlid might be parsed as xmlid only with module name. 🤔

Reply:
If no prefix is specified, the default module name will be `__import__`: https://github.com/odoo/odoo/blob/f29a674953c8631553c09ec4eb93bc4aca0849e5/odoo/models.py#L984 . If you specify the xml id `test` in the import file, it will get converted to `__import__.test` here: https://github.com/odoo/odoo/blob/f29a674953c8631553c09ec4eb93bc4aca0849e5/odoo/models.py#L1109 .